crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -1189,6 +1194,110 @@ impl Drop for IndexeddbCryptoStore {
     }
 }
 
+/// Fetch items from an object store in batches, transform each item using
+/// the supplied function, and stuff the transformed items into a single
+/// vector to return.
+async fn fetch_from_object_store_batched<R, F>(
+    object_store: IdbObjectStore<'_>,
+    f: F,
+    batch_size: usize,
+) -> Result<Vec<R>>
+where
+    F: Fn(JsValue) -> Result<R>,
+{
+    enum NextBatch {
+        First,          // In the first batch we don't start from a key
+        Later(JsValue), // In later batches we have a starting key
+    }
+
+    let mut result = Vec::new();
+    let mut next_batch = NextBatch::First;
+    let mut batch_n = 0;
+
+    loop {
+        debug!("Fetching Indexed DB records starting from {}", batch_n * batch_size);
+
+        // See https://github.com/Alorel/rust-indexed-db/issues/31 - we
+        // would like to use `get_all_with_key_and_limit` if it ever exists
+        // but for now we use a cursor and manually limit batch size.
+
+        // Get hold of a cursor for this batch
+        let cursor = match &next_batch {
+            NextBatch::First => object_store.open_cursor(),
+
+            NextBatch::Later(latest_key) => {
+                // We want a cursor starting after the last key we processed
+                let after_latest_key = IdbKeyRange::lower_bound_with_open(latest_key, true);
+
+                // This key should always be valid since we got hold of it
+                // within the same transaction.
+                let after_latest_key = after_latest_key.expect("Key was not valid!");
+
+                object_store.open_cursor_with_range(&after_latest_key)
+            }
+        }?
+        .await?;
+
+        // Fetch batch_size records into result
+        let next_key = fetch_batch(cursor, batch_size, &f, &mut result).await?;
+        if let Some(next_key) = next_key {
+            next_batch = NextBatch::Later(next_key);
+        } else {
+            break;
+        }
+
+        batch_n += 1;
+    }
+
+    Ok(result)
+}
+
+/// Fetch batch_size records from the supplied cursor,
+/// and return the last key we processed, or None if
+/// we reached the end of the cursor.
+async fn fetch_batch<R, F, Q>(
+    cursor: Option<IdbCursorWithValue<'_, Q>>,
+    batch_size: usize,
+    f: &F,
+    result: &mut Vec<R>,
+) -> Result<Option<JsValue>>
+where
+    F: Fn(JsValue) -> Result<R>,
+    Q: IdbQuerySource,
+{
+    let Some(cursor) = cursor else {
+        // Cursor was None - there are no more records
+        return Ok(None);
+    };
+
+    let mut latest_key = None;
+
+    for _ in 0..batch_size {
+        // Process the record
+        let processed = f(cursor.value());
```

Comment:
Would it make sense to run the processing outside the scope of the transaction (so first read with the transaction, then drop the transaction and process), so the table is available as early as possible by other tasks?

Reply:
Yes, that would make sense, but would it be OK to delay this change until a later PR where I split into multiple transactions? I'd prefer to ship this one soon and get feedback on whether it fixes the bug.

Reply:
Sure